File: graphs/breadth_first_search.go
package main

import (
	"fmt"
)

type Graph struct {
	adj map[int][]int
}

func NewGraph() *Graph {
	return &Graph{
		adj: make(map[int][]int),
	}
}

func (g *Graph) AddEdge(v, w int) {
	g.adj[v] = append(g.adj[v], w)
}

func (g *Graph) BFS(s int) {
	visited := make(map[int]bool)
	queue := []int{s}

	visited[s] = true

	for len(queue) > 0 {
		u := queue[0]
		queue = queue[1:]
		fmt.Printf("%d ", u)

		for _, v := range g.adj[u] {
			if !visited[v] {
				visited[v] = true
				queue = append(queue, v)
			}
		}
	}
	fmt.Println()
}

func main() {
	g := NewGraph()
	g.AddEdge(0, 1)
	g.AddEdge(0, 2)
	g.AddEdge(1, 2)
	g.AddEdge(2, 0)
	g.AddEdge(2, 3)
	g.AddEdge(3, 3)

	fmt.Println("Following is Breadth First Traversal (starting from vertex 2)")
	g.BFS(2)
}

File: graphs/breadth_first_search.js
/**
 * Breadth First Search Algorithm
 */

class Graph {
    constructor() {
        this.adj = new Map();
    }

    addEdge(v, w) {
        if (!this.adj.has(v)) {
            this.adj.set(v, []);
        }
        this.adj.get(v).push(w);
    }

    BFS(s) {
        let visited = new Set();
        let queue = [];

        visited.add(s);
        queue.push(s);

        while (queue.length > 0) {
            s = queue.shift();
            console.log(s + " ");

            let neighbors = this.adj.get(s);
            if (neighbors) {
                for (let i of neighbors) {
                    if (!visited.has(i)) {
                        visited.add(i);
                        queue.push(i);
                    }
                }
            }
        }
    }
}

// Driver Code
let g = new Graph();
g.addEdge(0, 1);
g.addEdge(0, 2);
g.addEdge(1, 2);
g.addEdge(2, 0);
g.addEdge(2, 3);
g.addEdge(3, 3);

console.log("Following is Breadth First Traversal (starting from vertex 2)");
g.BFS(2);